LCLSGeneral/LCLSGeneral/POUs/Data/FB_EpicsCentroidMonitor.TcPOU

```diff
@@ -0,0 +1,223 @@
+    threshold, says that if the centroid value doesn't change by at least this
+    on a frame-to-frame basis, consider this a result of a faulty AreaDetector
+    implementation.  The monitor would report values below this threshold as
+    "not updating" - ``bIsUpdating`` as ``FALSE``.
+
+    What needs to happen to get new centroid values?
+
+    - An updated array count
+    - An updated X centroid value (above fNewFrameMinimumChange)
+    - An updated Y centroid value (above fNewFrameMinimumChange)
+    - NO_ALARM severities for all values
+
+    This will result in an updated:
+
+    - fFrameTime (seconds between valid frames)
+
+    What needs to happen for ``bIsUpdating`` to be TRUE?
+
+    - All items from the above "new centroid values"
+    - Frame time below fMaximumFrameTime
+    - Centroid X and Y values above fMinimumValidChange
+
+*)
+FUNCTION_BLOCK FB_EpicsCentroidMonitor
+VAR_INPUT
+    fMaximumFrameTime : LREAL := 0.2;
+    (* Minimum change to be considered updating correctly - buggy detectors may need this. *)
+    fMinimumValidChange : LREAL := 1E-6;
+    (* Minimum change in pixels to be considered a new value for X, Y *)
+    fNewFrameMinimumChange : LREAL := 1E-6;
+    bEnable : BOOL := TRUE;
+END_VAR
+VAR_OUTPUT
+    {attribute 'pytmc' := '
+        pv: IsUpdating
+        io: input
+    '}
+    bIsUpdating : BOOL;
+    {attribute 'pytmc' := '
+        pv: CentroidX
+        io: input
+    '}
+    fCentroidX : LREAL;
+    {attribute 'pytmc' := '
+        pv: CentroidY
+        io: input
+    '}
+    fCentroidY : LREAL;
+    {attribute 'pytmc' := '
+        pv: ArrayCount
+        io: input
+    '}
+    nArrayCount : UDINT;
+    bValid : BOOL;
+    {attribute 'pytmc' := '
+        pv: FrameTime
+        io: input
+        field: DESC Time between frame updates
+        field: EGU sec
+    '}
+    fFrameTime : LREAL;
+END_VAR
+VAR
+    {attribute 'pytmc' := '
+        pv: CX_
+        link: CentroidX_RBV
+    '}
+    fbCentroidX : FB_LREALFromEPICS;
+
+    {attribute 'pytmc' := '
+        pv: CY_
+        link: CentroidY_RBV
+    '}
+    fbCentroidY : FB_LREALFromEPICS;
+
+    fLastCentroidX : LREAL;
+    fLastCentroidY : LREAL;
+
+    {attribute 'pytmc' := '
+        pv: Cnt_
+        link: ArrayCounter_RBV
+    '}
+    fbArrayCounter : FB_LREALFromEPICS;
+
+    // Last array count value
+    nLastArrayCount : UDINT;
+    // Time of the last frame update
+    tLastUpdate: TIME;
+
+    bInit : BOOL;
+    // Did we see a frame update yet? (FALSE at startup; TRUE after first frame.)
+    bSawFrame : BOOL;
+    // Do we have a new frame? Has the array count updated, and position change above fNewFrameMinimumChange?
+    bHaveNewFrame : BOOL;
+    // For this new frame, is it above the threshold fMinimumValidChange?
+    bAboveThreshold : BOOL;
+
+END_VAR]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[IF NOT bEnable THEN
+    bValid := FALSE;
+    RETURN;
+END_IF
```

Comment:
This is much better than what I've been doing with bEnable (wrapping the entire FB in an IF statement)